Frameworks/AddressBook/ABRecord.mm

```diff
@@ -20,12 +20,82 @@
 #import "ABContactInternal.h"
 #import "ABRecordInternal.h"
 #import "ABSourceInternal.h"
 #import "ABMultiValueInternal.h"
 #import "NSDate+AddressBookAdditions.h"
 
+// Verifies that the given value doesn't have a length greater than the given length. If value's length
+// is longer, returns false and, if error is not null, creates an error with the given propertyName.
+// Otherwise, returns true and does nothing with error.
+static bool _checkLength(NSString* value, NSUInteger length, CFErrorRef* error, NSString* propertyName) {
+    if ([value length] > length) {
+        if (error) {
+            NSString* reason = [NSString stringWithFormat:@"The %@ property must not be more than %d characters\n", propertyName, length];
+            NSDictionary* userInfo = @{
+                NSLocalizedDescriptionKey : NSLocalizedString(@"Error setting record value.\n", nil),
+                NSLocalizedFailureReasonErrorKey : NSLocalizedString(reason, nil)
+            };
+            *error = (__bridge_retained CFErrorRef)[NSError errorWithDomain:(__bridge NSString*)ABAddressBookErrorDomain
+                                                                       code:kABOperationNotPermittedByStoreError
+                                                                   userInfo:userInfo];
+        }
+
+        return false;
+    }
+
+    return true;
+}
+
+// Ensures that the property type of the multiValue matches the expected ABPropertyType. If they don't
+// match, returns false and, if error is not null, creates an error describing the problem.
+// Otherwise, returns true and does nothing with error.
+static bool _checkType(ABPropertyType expected, _ABMultiValue* multiValue, CFErrorRef* error) {
+    if (expected != [multiValue getPropertyType]) {
+        if (error) {
+            NSDictionary* userInfo = @{
+                NSLocalizedDescriptionKey : NSLocalizedString(@"Error setting record value.\n", nil),
+                NSLocalizedFailureReasonErrorKey : NSLocalizedString(@"The property type of the multi-value was incorrect!\n", nil)
+            };
+            *error = (__bridge_retained CFErrorRef)[NSError errorWithDomain:(__bridge NSString*)ABAddressBookErrorDomain
+                                                                       code:kABOperationNotPermittedByStoreError
+                                                                   userInfo:userInfo];
+        }
+
+        return false;
+    }
+
+    return true;
+}
+
+// Updates the address book manager if a read-write
+// contact is modified to ensure changes persist if
+// the address book is saved.
+static void _updateManager(_ABContact* person) {
+    if (person.type == kAddressBookReadWriteContact) {
+        [(__bridge _ABAddressBookManager*)person.manager modifyContact:(__bridge ABRecordRef)person];
+    }
+}
+
+// The maximum allowed String length of various Windows Contacts properties.
+static const NSUInteger kABPersonFirstNameLength = 64;
```

Comment:
Is there documentation of this? It might be helpful to link to where you got these numbers so anybody looking at this can understand why these limits are set.